agent/unit/graph_test.go

```diff
@@ -0,0 +1,645 @@
+		const numGoroutines = 100
+		const edgesPerGoroutine = 10
+
+		// Launch goroutines to add edges concurrently
+		errors := make([]error, numGoroutines*edgesPerGoroutine)
+		errorIndex := 0
+		var mu sync.Mutex
+
+		for i := 0; i < numGoroutines; i++ {
+			wg.Add(1)
+			go func(goroutineID int) {
+				defer wg.Done()
+				for j := 0; j < edgesPerGoroutine; j++ {
+					from := &Unit{Name: fmt.Sprintf("unit-%d-%d", goroutineID, j), Status: StatusPending}
+					to := &Unit{Name: fmt.Sprintf("unit-%d-%d", goroutineID, j+1), Status: StatusPending}
+					err := graph.AddEdge(from, to, StatusCompleted)
+
+					mu.Lock()
+					errors[errorIndex] = err
+					errorIndex++
+					mu.Unlock()
+				}
+			}(i)
+		}
+
+		wg.Wait()
+
+		// Verify no errors occurred during concurrent operations
+		for i, err := range errors {
+			require.NoError(t, err, "error at index %d", i)
+		}
+
+		// Save DOT file for analysis
+		saveDOTFile(t, graph, "concurrent-add-edge")
+
+		// Verify no panics occurred and graph is in valid state
+		// We can't easily count total edges due to concurrent access, but we can verify
+		// the graph is still functional
+		dot, err := graph.ToDOT("test")
+		require.NoError(t, err)
+		require.NotEmpty(t, dot)
+	})
+
+	t.Run("ConcurrentReads", func(t *testing.T) {
+		t.Parallel()
+
+		graph := &unit.Graph[Status, *Unit]{}
+
+		// Pre-populate graph with 50 vertices and edges
+		units := make([]*Unit, 50)
+		for i := 0; i < 50; i++ {
+			units[i] = &Unit{Name: fmt.Sprintf("unit-%d", i), Status: StatusPending}
+		}
+
+		// Add edges to create a chain
+		for i := 0; i < 49; i++ {
+			err := graph.AddEdge(units[i], units[i+1], StatusCompleted)
+			require.NoError(t, err)
+		}
+
+		// Save initial DOT file
+		saveDOTFile(t, graph, "concurrent-reads-initial")
+
+		var wg sync.WaitGroup
+		const numReaders = 200
+
+		// Launch readers
+		results := make([]struct {
+			forwardLen int
+			reverseLen int
+			panicked   bool
+		}, numReaders)
+
+		for i := 0; i < numReaders; i++ {
+			wg.Add(1)
+			go func(readerID int) {
+				defer wg.Done()
+				defer func() {
+					if r := recover(); r != nil {
+						results[readerID].panicked = true
+					}
+				}()
+
+				// Read from random vertices
+				vertex := units[secureRandInt(len(units))]
+				forwardEdges := graph.GetForwardAdjacentVertices(vertex)
+				reverseEdges := graph.GetReverseAdjacentVertices(vertex)
+
+				// Store results for verification outside goroutine
+				results[readerID].forwardLen = len(forwardEdges)
+				results[readerID].reverseLen = len(reverseEdges)
+			}(i)
+		}
+
+		wg.Wait()
+
+		// Verify results outside of goroutines
+		for i, result := range results {
+			require.False(t, result.panicked, "reader %d panicked", i)
+			require.LessOrEqual(t, result.forwardLen, 1, "reader %d: chain has max 1 forward edge", i)
+			require.LessOrEqual(t, result.reverseLen, 1, "reader %d: chain has max 1 reverse edge", i)
+		}
+	})
+
+	t.Run("ConcurrentReadWrite", func(t *testing.T) {
```

Comment:
I think we could potentially just keep this concurrency test and drop the others.